### ISSUE TITLE
Implement a simple Kirchoff law for Extensive port in pyomo.network

### ISSUE THREAD

**Issue by ReinboldV** [Pyomo/pyomo#882]
Hello,
I am experimenting the pyomo network for hierarchical modelling. First, thanks a lot for your contributions. This is a very useful tool !
I have some hard time understanding the philosophy behind the Port.Extensive rule when more than one arc is expanded.

Let's take a simple example with 3 blocks, one "source", b1, connected to 2 "destinations", b2 and b3. Let's assume that each destination is connected to the source through a port that expose an extensive variable, i.e. a flow.

Here a simple code to do this : 

```
from pyomo.environ import *
from pyomo.network import *
from pyomo.dae import *

m = ConcreteModel()

m.b1 = Block()
m.b2 = Block()
m.b3 = Block()

def source_block(b):
    b.t = ContinuousSet(bounds=(0, 1))
    b.p_out = Var(b.t)
    b.outlet = Port(initialize={'p': (b.p_out, Port.Extensive)})

def load_block(b):
    b.t = ContinuousSet(bounds=(0, 1))
    b.p_in = Var(b.t)
    b.inlet = Port(initialize={'p': (b.p_in, Port.Extensive)})

source_block(m.b1)
load_block(m.b2)
load_block(m.b3)

m.b12 = Arc(source=m.b1.outlet, destination=m.b2.inlet)
m.b13 = Arc(source=m.b1.outlet, destination=m.b3.inlet)

TransformationFactory('dae.finite_difference').apply_to(m, nfe=1)
TransformationFactory("network.expand_arcs").apply_to(m)

m.pprint()
```
I don't really understand why one would introduce the following variables : `b13_expanded.splitfact` and `b13_expanded.splitfact` the following constraints `m.b12_expanded.p_split` and `m.b13_expanded.p_split` within the new connection blocks. In a general flow  connection, one would only need a Kirchoff law (i.e. sum of flows are equal to 0) without imposing split (the physic will actually take care of this.), i.e. : `b1.p_out[i] == b2.p_in[t] + b3.pin[t]` which is equivalent to the set of generated constraints `b1.outlet_p_outsum`,  `b2.inlet_p_insum` and `b3.inlet_p_insum`. So here is my questions/remarks : 

- Would this Kirchoff law be a new kind of expansion for arcs ?
- Is there an easy way to automatically disable the constraints `b12_expanded.p_split` and `b13_expanded.p_split` ? Or make this behaviour an option ? In the call for Arc for instance.

Thanks again for your contribution and for your help, 
Greetings

**Comment by gseastream** [comment]
If I understand correctly, I believe the behavior you're seeing is necessary for the expansion to be correct. If you were to skip the split fractions and just write an indexed constraint for all of the `p` variables saying the source `p` equals the sum of the dest `p`'s, you would achieve the physical property where the flow is conserved, but since `p` is an indexed variable (in this case over time), there would be no guarantee that each arc contains an equal "portion" of `p` at every time index.

Consider a variable `x` indexed over the set `{1, 2}`. Without `splitfrac` variables, we would get the constraints

```
b1.x[1] == b2.x[1] + b3.x[1]
b1.x[2] == b2.x[2] + b3.x[2]
```

If for example we fix `b1.x[1] = b1.x[2] = 5`, then these constraints allow for the following values

```
b2.x[1] = 3
b3.x[1] = 2
b2.x[2] = 1
b2.x[2] = 4
```

This violates the intention of the extensive rule, since we basically have `x` being split in different portions for its different indexes. That was why split fractions were ([almost](https://pyomo.readthedocs.io/en/latest/modeling_extensions/network.html#pyomo.network.Port.Extensive)) always included, since using them across all of the `_split` constraints guarantees that each destination arc contains an equal portion of every variable entering it.

**Comment by ReinboldV** [comment]
Hi, thanks for your answer and explanations !
I see your point in introducing this fraction variable and I am sure this allows to deal with real applications in graph optimization. However, I believe this is a particular case. 

For my dynamic physical graph applications, I am looking for a 'classical' flow conservation constraint, without assuming constant split (exactly as you described). This is actually more generic in my perspective. In thermodynamics for instance, the heat-flow variables are deduced from effort variables, or potentials (Temperature in this case). 
As a consequence, considering fixed split fraction would be equivalent to fixing temperature differences for all instant. Fixing this split might be a nice simplification in some cases. But it is a strong hypothesis in the general case, in my point of view. In definitive, I'd like to let ``x` being split in different portions for its different index. 
This behaviour is the same for a large variety of applications : thermal (temperature / heat-flow), fluids (pressure / flow), electric (voltage / current), mechanic (force / velocity) . This kind of graph are known as 'bond-graph'. 

In conclusion, I believe that a "Conservative" rule might be a good enhancement for the network module. What is your position on this ? In your opinion, is this new rule, lets call it Conservative, would be hard to develop ? Could it be a new PR ? 

Greetings

**Comment by gseastream** [comment]
I agree that what you're looking for differs from the purpose that the Extensive rule serves. The desired behavior could be implemented in one of at least two ways. One would be to create a new rule, like Conservative, which could write simpler summation constraints for the variables that get created on the arcs' expanded blocks (referred to as `evar`'s in the code). However, in order to implement the use of "effort variables" in your example, you would (I assume) need to write more than just summations. Moreover, all of your different applications you listed probably have different needs in regards to dictating how flows split, so different procedures would need to be implemented for them.

You could either implement the use of effort variables (or whatever your application needs) within the Conservative rule procedure, having it generate both summations as well as constraints that specify the specific flows, or you could implement that within your block that contains the outlet port. Yet, if you were to implement such constraints on the block, you would need to do so after expanding the arcs since you would presumably need access to the evars which only get created during expansion. Alternatively, you could also break down the complexity further by giving your block multiple outlet ports which contain their own variables, and which are 1-to-1 connected to their downstream ports, and then place constraints on your block that specify the variables in those ports so it basically does the splitting itself without using the arc expander rules. Of course, this option eliminates the convenience of the implicit mixing and splitting rules.

In summary, you can get your desired behavior with a new rule, but this rule would also be a good place to add the use of effort variables or whatever physical attributes dictate flow splits at a given instant. But you might need a different rule for every different relevant application. And if not within that rule, these splitting constraints could be written on the block but it might not be as clean.

The expansion rule framework was left open to allow for the use of custom rules, or a PR can be opened to add the rule to the Port class for wider use. You can look at the code for Equality and Extensive [here](https://github.com/Pyomo/pyomo/blob/6c042537da93146eaef1398aa544bda0e5f85a56/pyomo/network/port.py#L443), and hopefully they are well commented enough to see what went into them and give a shot at writing your own.

**Comment by ReinboldV** [comment]
Hi Grant,  

Thanks a lot for your answer !
I am not a python expert, but I'll give a try to write my own rule and hopefully add it as a PR.
I guess I only have to follow the Extensive rule. 
Thanks for you guidance.
